### ISSUE TITLE
Problem running async Actions

### ISSUE THREAD

**Issue by misirlou-tg** [dotnet/command-line-api#2152]
While porting an existing project from the beta4 package in nuget to a more recent build (beta4.23181.5) I saw that async actions were not running to completion. After some debugging I found that if I made the action non-async, it would work as expected. I was able to reproduce it with the following.
```c#
using System;
using System.CommandLine;
using System.Threading.Tasks;

namespace CliAsyncTest;

public class Program
{
    static async Task<int> Main(string[] args)
    {
        var rootCmd = new CliRootCommand("CliAsyncTest - Test InvokeAsync()");
        rootCmd.SetAction(async (_) =>  // Fix: remove async
        {
            Console.WriteLine("rootCmd Action starting");
            await Task.Delay(100);  // Fix: remove await, add .Wait()
            Console.WriteLine("rootCmd Action complete");
        });

        return await rootCmd.Parse(args).InvokeAsync();
    }
}
```
When I run this I will see the first line output to the console, but not the second. After applying the fixes in the comments I will see both lines output.

Tested with:
- beta4.23181.5
- Windows 10

**Comment by elgonzo** [comment]
Not a bug (technically), just your lambda function declaration and the allowed implicit lambda-to-delegate conversions conspiring to choose an undesired SetAction method overload.

Note that the async command actions accepted by the SetAction overloads have two input parameters, a ParseResult and a CancellationToken. However, your anonymous lambda function specifies only a single input parameter (a discard). From all the SetAction overloads, this is the closest matching one: 

https://github.com/dotnet/command-line-api/blob/2819f48c9520f1de5ce2228bccf5f8eb7d39e881/src/System.CommandLine/CliCommand.cs#L103-L106

Why this precisely? Because of all the different SetAction overloads with all their different Action/Func/delegate parameters, only Action\<object\> is a delegate type your lambda function can be converted into (because your lambda has only a single input parameter, and it is permissible to ignore the return value).

And because your code happens to use a SetAction overload that is meant for synchronous action delegates, you will experience the observed effect.

Solution: Make sure the signature of your lambda function matches the delegate type of one of the SetAction overloads meant for async actions, e.g.:
```
rootCmd.SetAction(async (_, cancellationToken) =>  // Fix: remove async
{
    ...
}
```

Still, this might indicate a problem with the SetAction API as it allows making these kind of mistakes easy.

An idea: To reduce the likelihood of this kind of confusion/problem happening, perhaps the `SetAction` method overloads for async actions could be renamed to `SetAsyncAction` or something similar. If for example the erroneous code would attempt `rootCmd.SetAsyncAction(async (_) => ...);` then this would result in a useful compiler error instead of unintentionally picking an ill-fitting synchronous SetAction method overload.

**Comment by misirlou-tg** [comment]
Yes, that fixed it for me. I agree it is easy to make this mistake. The code was previously using `SetHandler()` and it did not have a param in the lambda for the cancellation token, but it did work.

**Comment by jonsequitur** [comment]
This is very much a bug because it's easy to get it wrong. We shouldn't require people to implement (including via `SetAction` delegates) both `Invoke` and `InvokeAsync`. Calling `Invoke` or `InvokeAsync` on the `ParseResult` should be equivalent in terms of your app logic without you having to do extra work to accommodate both cases.
